src/waveform/renderers/allshader/digitsrenderer.cpp

```diff
@@ -0,0 +1,226 @@
+#include "waveform/renderers/allshader/digitsrenderer.h"
+
+#include <QColor>
+#include <QFontMetricsF>
+#include <QGraphicsBlurEffect>
+#include <QGraphicsPixmapItem>
+#include <QGraphicsScene>
+#include <QOpenGLTexture>
+#include <QPainter>
+#include <QPainterPath>
+
+#include "./util/assert.h"
+#include "waveform/renderers/allshader/matrixforwidgetgeometry.h"
+#include "waveform/renderers/allshader/vertexdata.h"
+
+// Render digits using a texture (generated) with digits with blurred dark outline
+
+namespace {
+
+// The texture will contain 12 characters: 10 digits, colon and dot
+constexpr int NUM_CHARS = 12;
+
+// space around chars for blurred dark outline
+constexpr qreal SPACE = 3;
+// extra kerning when rendering chars
+constexpr qreal KERNING = 1;
+
+constexpr char indexToChar(int index) {
+    constexpr char str[] = "0123456789:.";
+    return str[index];
+}
+constexpr int charToIndex(QChar ch) {
+    int value = ch.toLatin1() - '0';
+    if (value >= 0 && value <= 9) {
+        return value;
+    }
+    if (ch == ':') {
+        return 10;
+    }
+    if (ch == '.') {
+        return 11;
+    }
+    DEBUG_ASSERT(false);
+    return 11; // fallback to dot
+}
+} // namespace
+
+allshader::DigitsRenderer::~DigitsRenderer() = default;
+
+void allshader::DigitsRenderer::init() {
+    initializeOpenGLFunctions();
+    m_shader.init();
+}
+
+float allshader::DigitsRenderer::height() const {
+    return m_height;
+}
+
+void allshader::DigitsRenderer::updateTexture(
+        float fontPointSize, float maxHeight, float devicePixelRatio) {
+    // Check if we need to update the texture: when the requested font size
+    // changes, or when the max height is too small for the current font size or
+    // different from the max height we previously adjusted the font size for.
+    if (fontPointSize == m_fontPointSize && maxHeight >= m_height &&
+            (m_adjustedForMaxHeight == 0.f ||
+                    maxHeight == m_adjustedForMaxHeight)) {
+        return;
+    }
+
+    m_fontPointSize = fontPointSize;
+    m_adjustedForMaxHeight = 0.f;
+
+    QFont font;
+    QFontMetricsF metrics{font};
+    font.setFamily("Open Sans");
+    qreal totalTextWidth;
+    qreal maxTextHeight;
+    qreal heightWithPadding;
+
+    bool retry = false;
+    do {
+        font.setPointSizeF(fontPointSize);
+
+        metrics = QFontMetricsF{font};
+
+        totalTextWidth = 0;
+        maxTextHeight = 0;
+
+        for (int i = 0; i < NUM_CHARS; i++) {
+            assert(charToIndex(indexToChar(i)) == i);
```

Comment:
We avoid plain asserts, because they may crash in release builds. 

You may put this into the anonymous namespace instead: 
```
constexpr bool checkCharToIndex() {
    for (int i = 0; i < NUM_CHARS; i++) {
        if (charToIndex(indexToChar(i)) != i) {
        	return false;
        }
    }
    return true;
}
static_assert(checkCharToIndex());
```

Reply:
done